src/tools/pkcs11-tool.c

```diff
@@ -1400,22 +1469,37 @@ int main(int argc, char * argv[])
 	}
 
 	if (do_sign || do_derive) {
-
 		/*
 		 * Newer mechanisms have their details in the mechanism table, however
 		 * if it's not known fall back to the old code always assuming it was a
 		 * CKO_PRIVATE_KEY.
 		 */
 		if (mf_flags != MF_UNKNOWN) {
 			/* this function dies on error via util_fatal */
-			find_object_flags(session, mf_flags, &object,
-				opt_object_id_len ? opt_object_id : NULL,
-				opt_object_id_len, 0);
+			if (opt_uri) {
+				find_object_flags_and_type_or_id_or_label(session, mf_flags, &object,
```

Comment:
can we do the search here directly by the URI, instead of limiting only to class, id, label?